### PR TITLE
fix(shell): bound denylist matches to argument tokens

### LINKED ISSUE #3757: fix(shell): denylist prefix-matches absolute cleanup paths and --allow-empty
## Summary
The shell denylist matches prefixes rather than complete arguments. An explicit temporary-directory cleanup is rejected as root deletion; similarly, `--allow-empty` is rejected as `--all`.

Verified against commit 21125a6144b7e32a73659cae8994f4185a17a946, using the installed editable package.

## Safe reproduction (no shell commands executed)
```python
from gptme.tools.shell_validation import is_denylisted
for command in [
    "rm -rf /tmp/gptme-fp-example",
    "rm -rf /",
    "git commit --allow-empty -m x",
]:
    denied, reason, matched = is_denylisted(command)
    print(command, denied, repr(matched))
```
Actual:
```text
rm -rf /tmp/gptme-fp-example True 'rm -rf /'
rm -rf / True 'rm -rf /'
git commit --allow-empty -m x True 'git commit --all'
```

## Cause / expected behavior
In `gptme/tools/shell_validation.py`, `re.search` applies patterns including `rm\s+-rf\s+/` and `git\s+commit\s+--all` without an argument-end boundary.

Distinguish an exact root operand from a longer absolute path, and an exact option from a longer option. An explicit tmp path should not be classified as deletion of root (normal confirmation/policy may still apply). Preserve root-deletion protection and add regression tests for argument boundaries, quoting, and shell separators.

The allow-empty finding was also mentioned in #3756, but is independent of that report's unverified hang diagnosis.

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

The shell denylist used unanchored prefix matches, so:

- `rm -rf /tmp/gptme-fp-example` was denied as `rm -rf /`
- `git commit --allow-empty` was denied as `git commit --all`

This keeps true root-deletion and bulk-commit flags blocked, but requires operand/option boundaries (including separators and quoted root operands).

`pkill`/`killall` policy is unchanged.

Fixes #3757

## Test plan

- [x] `uv run pytest tests/test_tools_shell_validation.py` (439 passed)
- [x] Added regressions for `/tmp` cleanup, quoted `/`, separators, `--allow-empty`, `--amend`, and `-am`

### CODE REVIEW

File: gptme/tools/shell_validation.py [hunk 1]
```diff
@@ -85,10 +85,10 @@
...
         [
             r"git\s+add\s+\.(?:\s|$)",  # Match 'git add .' but not '.gitignore'
-            r"git\s+add\s+-A",
-            r"git\s+add\s+--all",
-            r"git\s+commit\s+-a",
-            r"git\s+commit\s+--all",
+            r"git\s+add\s+-A(?:\s|$)",
+            r"git\s+add\s+--all(?:\s|$)",
+            r"git\s+commit\s+-[a-zA-Z]*a[a-zA-Z]*(?:\s|$)",  # -a / -am / -ma, not --amend
+            r"git\s+commit\s+--all(?:\s|$)",
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/shell_validation.py]: Fixed in 7a4e120de. Git denylist token boundaries now include `;`, `&`, and `|`, so `git add -A; git commit`, `git commit --all&&git push`, and `git commit -am|cat` match.

File: gptme/tools/shell_validation.py [hunk 2]
```diff
@@ -85,10 +85,10 @@
     (
         [
             r"git\s+add\s+\.(?:\s|$)",  # Match 'git add .' but not '.gitignore'
-            r"git\s+add\s+-A",
-            r"git\s+add\s+--all",
-            r"git\s+commit\s+-a",
-            r"git\s+commit\s+--all",
+            r"git\s+add\s+-A(?:\s|$)",
+            r"git\s+add\s+--all(?:\s|$)",
+            r"git\s+commit\s+-[a-zA-Z]*a[a-zA-Z]*(?:\s|$)",  # -a / -am / -ma, not --amend
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/shell_validation.py]: Fixed in 7a4e120de. `-a` must lead the short-option cluster (`-a` / `-am`), so `-ma` and `-mabc` (message values) are no longer treated as bulk commits.

File: gptme/tools/shell_validation.py [hunk 3]
```diff
@@ -104,8 +104,10 @@
     ),
     (
         [
-            r"rm\s+-rf\s+/",
-            r"sudo\s+rm\s+-rf\s+/",
+            # Root operand only: `/`, `/.`, `/..`, `//`, `/*`, optionally quoted.
+            # Do not prefix-match longer absolute paths such as `/tmp/foo`.
+            r"rm\s+-rf\s+['\"]?/(?:(?:\.){1,2}|/|\*)?['\"]?(?:\s|$|[;&|])",
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/shell_validation.py]: Fixed in 7a4e120de. Root matching now accepts repeated `/`, `.`, `*`, and quotes, so `/./`, `/..`, `/""`, and `/''` are denied without prefix-matching `/tmp/...`.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
## 🤖 AI code review

This PR refines shell validation by improving command name allowlisting, adding proper quote and heredoc parsing, strengthening pipe and redirection detection, and preventing path traversal attacks. It also expands the test suite to cover these scenarios.

### Safe to merge — no P0/P1 findings

Confidence 5/5

✅ **No findings.** The diff looks correct to me on this pass.

<details><summary><b>Files changed (2)</b> — the diff as I read it</summary>

- `gptme/tools/shell_validation.py` — Fixes to shell validation logic, including allowlist/denylist checks, quote/heredoc parsing, pipe detection, redirection handling, and path traversal protection.
- `tests/test_tools_shell_validation.py` — Expanded test coverage for the shell validation edge cases addressed in the fixes.

</details>

<details><summary>Previous review passes</summary>

| commit | score | findings | engine | when |
| --- | --- | --- | --- | --- |
| `7a4e120deaea` | 5/5 | 0 | llm | 2026-09-08 12:20 UTC |
| `f654ff7f1c07` | 5/5 | 0 | llm | 2026-09-08 12:42 UTC |
| `e0906cea193d` | 3/5 | 2 | llm | 2026-09-08 13:02 UTC |

</details>

<sub>Reviewed `e0906cea193d` · `openrouter/deepseek/deepseek-v4-flash-0731` · `llm` engine · 245s · [about this reviewer](https://github.com/gptme/gptme-contrib/blob/master/docs/ai-reviewer.md)</sub>

<details><summary><sub>Maintainer commands</sub></summary>

<sub>`@TimeToBuildBob review` (own line) — fresh review · `@TimeToBuildBob fix` — a worker acts on the findings. Once per comment; 👀 = received.</sub>

</details>

File: gptme/tools/shell_validation.py [hunk 4]
```diff
@@ -104,8 +107,11 @@
     ),
     (
         [
-            r"rm\s+-rf\s+/",
-            r"sudo\s+rm\s+-rf\s+/",
+            # Root-equivalent operand: `/`, `/./`, `/..`, `//`, `/*`, `'/'`,
+            # `/""`. Quotes and `.`/`*` after `/` may repeat; a real path
+            # component such as `/tmp/foo` does not match.
+            r"rm\s+-rf\s+['\"]*(?:/(?:\.{1,2}|\*)?)+['\"]*(?:\s|$|[;&|])",
+            r"sudo\s+rm\s+-rf\s+['\"]*(?:/(?:\.{1,2}|\*)?)+['\"]*(?:\s|$|[;&|])",
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/shell_validation.py]: Fixed in f654ff7f1. Token boundaries now include `)`, so `(rm -rf /)`, `$(sudo rm -rf /)`, `(git add -A)`, and `$(git commit --all)` match. Grouped `/tmp/...` still allowed.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/tools/shell_validation.py [hunk 5]
```diff
@@ -104,8 +108,11 @@
     ),
     (
         [
-            r"rm\s+-rf\s+/",
-            r"sudo\s+rm\s+-rf\s+/",
+            # Root-equivalent operand: `/`, `/./`, `/..`, `//`, `/*`, `'/'`,
+            # `/""`. Quotes and `.`/`*` after `/` may repeat; a real path
+            # component such as `/tmp/foo` does not match.
+            r"rm\s+-rf\s+['\"]*(?:/(?:\.{1,2}|\*)?)+['\"]*(?:\s|$|[;&|)])",
+            r"sudo\s+rm\s+-rf\s+['\"]*(?:/(?:\.{1,2}|\*)?)+['\"]*(?:\s|$|[;&|)])",
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/shell_validation.py]: Fixed in e0906cea1. `is_denylisted()` now walks every match, so quoted data such as `printf '%s\n' 'git commit --all)'` and `echo '(rm -rf /)'` is skipped, while a later unquoted command still matches. A trailing `)` is a grouping delimiter only when unquoted. Grouped `(rm -rf /)` / `$(git commit --all)` still denied.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by ErikBjare** [comment]
@TimeToBuildBob review

**Comment by TimeToBuildBob** [comment]
@ErikBjare Review on `e0906cea1`: AI reviewer 5/5 (0 findings; the 13:02 3/5 pass was superseded at 13:14), Greptile 5/5, CI green, review threads clear.

Not self-merging. Consensus primary (`deepseek-v4-flash`) times out at 45s; fallback `llama-3.1-8b-instruct` then blocks at 2/5 on a generic complexity hedge (`git commit -am` / "new regexes"). `-am` is an intentional bulk-commit deny, already on master. The 6h memo window holds this head at 2/5, so I can't override it. Ready for squash-merge.

**Comment by TimeToBuildBob** [comment]
Landed as `65d888f` (squash-merge by Erik). Review request is complete.

Verified the gptme/gptme#3757 reproduction on that commit (no shell commands executed):

- `rm -rf /tmp/gptme-fp-example` allowed
- `git commit --allow-empty -m x` allowed
- `rm -rf /` and `git commit --all`/`-a`/`-am` still denied

Closing gptme/gptme#3757 after that symptom check. `tests/test_tools_shell_validation.py`: 447 passed.
